File: code/javascript/data_structures_article-10.js
/**
 * Basics of graph theory algorithms
 */
export function solve(...args) {
  // Entry point for the course exercise: pass inputs directly.
  return allPathsSourceTarget(...args);
}

var allPathsSourceTarget = function(graph) {
    var res = [];
    var traverse = function(graph, s, path) {
        path.push(s);
        var n = graph.length;
        if (s === n - 1) {
            res.push(path.slice());
            path.pop();
            return;
        }
        for (var i = 0; i < graph[s].length; i++) {
            traverse(graph, graph[s][i], path);
        }
        path.pop();
    };
    var path = [];
    traverse(graph, 0, path);
    return res;
};

File: code/python/data_structures_article-10.py
"""Basics of graph theory algorithms"""

from __future__ import annotations
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Run this lesson exercise with positional inputs."""
    return _Solution().__init__(*args, **kwargs)


class _Solution:
    def __init__(self):
        self.res = []
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        path = []
        self.traverse(graph, 0, path)
        return self.res
    """  """
    def traverse(self, graph: List[List[int]], s: int, path: List[int]) -> None:
        path.append(s)
        n = len(graph)
        if s == n - 1:
            self.res.append(path[:])
            path.pop()
            return
        for v in graph[s]:
            self.traverse(graph, v, path)
        path.pop()
